File: v1/src/main/java/com/api/v1/medical_appointments/MedicalAppointment.java
package com.api.v1.medical_appointments;

import com.api.v1.customers.Customer;
import com.api.v1.doctors.Doctor;
import com.api.v1.medical_appointments.enums.MedicalAppointmentStatus;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.UUID;

@Document(collection = "MedicalAppointments")
public record MedicalAppointment(
        @Id
        String id,
        Customer customer,
        Doctor doctor,
        MedicalAppointmentStatus status,
        LocalDateTime createdAt,
        LocalDateTime bookedAt,
        LocalDateTime cancelledAt,
        LocalDateTime completedAt
) {

    public static MedicalAppointment of(Customer customer, Doctor doctor, LocalDateTime bookedAt) {
        return new MedicalAppointment(
                UUID.randomUUID().toString(),
                customer,
                doctor,
                MedicalAppointmentStatus.ACTIVE,
                LocalDateTime.now(),
                bookedAt,
                null,
                null
        );
    }
}

File: v2/src/main/kotlin/com/api/v2/medical_appointment/MedicalAppointment.kt
package com.api.v2.medical_appointment

import com.api.v2.customers.Customer
import com.api.v2.doctors.Doctor
import com.api.v2.medical_appointment.enums.MedicalAppointmentStatus
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDateTime
import java.util.UUID

@Document(collection = "MedicalAppointments")
data class MedicalAppointment(
    @Id
    val id: String,
    val customer: Customer,
    val doctor: Doctor,
    val status: MedicalAppointmentStatus,
    val createdAt: LocalDateTime,
    val bookedAt: LocalDateTime,
    val cancelledAt: LocalDateTime?,
    val completedAt: LocalDateTime?
) {

    companion object {
        fun of (customer: Customer,
                doctor: Doctor,
                bookedAt: LocalDateTime
        ): MedicalAppointment {
            return MedicalAppointment(
                UUID.randomUUID().toString(),
                customer,
                doctor,
                MedicalAppointmentStatus.ACTIVE,
                LocalDateTime.now(),
                bookedAt,
                null,
                null
            )
        }
    }

    fun markAsCancelled(): MedicalAppointment {
        return MedicalAppointment(
            this.id,
            this.customer,
            this.doctor,
            MedicalAppointmentStatus.CANCELLED,
            this.createdAt,
            this.bookedAt,
            LocalDateTime.now(),
            null
        )
    }

    fun markAsCompleted(): MedicalAppointment {
        return MedicalAppointment(
            this.id,
            this.customer,
            this.doctor,
            MedicalAppointmentStatus.COMPLETED,
            this.createdAt,
            this.bookedAt,
            null,
            LocalDateTime.now()
        )
    }
}